File: android/app/src/main/java/tv/swarmcast/p2p/SegmentStore.kt
package tv.swarmcast.p2p

import java.security.MessageDigest

class SegmentStore(private val maxBytes: Long = 64L * 1024 * 1024) {
    data class Entry(val seq: Int, val bytes: ByteArray, val sha256: String)

    private val map = LinkedHashMap<Int, Entry>(64, 0.75f, true)
    private var totalBytes = 0L

    @Synchronized fun get(seq: Int): Entry? = map[seq]

    @Synchronized fun heldSeqs(): Set<Int> = map.keys.toSet()

    @Synchronized fun putVerified(seq: Int, bytes: ByteArray, expectedSha256: String): Boolean {
        val actual = sha256(bytes)
        if (actual != expectedSha256) return false

        map[seq]?.let { totalBytes -= it.bytes.size }
        map.remove(seq)
        map[seq] = Entry(seq, bytes, actual)
        totalBytes += bytes.size

        val iterator = map.entries.iterator()
        while (totalBytes > maxBytes && iterator.hasNext()) {
            totalBytes -= iterator.next().value.bytes.size
            iterator.remove()
        }
        return true
    }

    private fun sha256(bytes: ByteArray): String =
        MessageDigest.getInstance("SHA-256")
            .digest(bytes)
            .joinToString("") { "%02x".format(it) }
}

File: packages/p2p/src/segmentStore.js
import { createHash } from "node:crypto";

export function sha256Hex(bytes) {
  return createHash("sha256").update(bytes).digest("hex");
}

export class SegmentStore {
  constructor({ maxBytes = 64 * 1024 * 1024 } = {}) {
    this.maxBytes = maxBytes;
    this.map = new Map();
    this.totalBytes = 0;
  }

  get(seq) {
    const entry = this.map.get(seq);
    if (!entry) return null;
    this.map.delete(seq);
    this.map.set(seq, entry);
    return entry;
  }

  heldSeqs() {
    return new Set(this.map.keys());
  }

  putVerified(seq, bytes, expectedSha256) {
    const body = Buffer.from(bytes);
    const actual = sha256Hex(body);
    if (actual !== expectedSha256) return false;

    const existing = this.map.get(seq);
    if (existing) {
      this.totalBytes -= existing.bytes.length;
      this.map.delete(seq);
    }

    const entry = { seq, bytes: body, sha256: actual };
    this.map.set(seq, entry);
    this.totalBytes += body.length;
    this.evict();
    return true;
  }

  evict() {
    while (this.totalBytes > this.maxBytes && this.map.size > 0) {
      const [seq, entry] = this.map.entries().next().value;
      this.map.delete(seq);
      this.totalBytes -= entry.bytes.length;
    }
  }
}
